### PR TITLE
drivers: use codegen for device instantiation

### PR DESCRIPTION
**Pull request by b0661**
# Overview

Use Codegen #10885 for device instantiation.

This PR includes parts of PR #10885. It is the successor to #9163.

# Motivation for or Use Case

Demonstrate the usage of Codegen and the included devicedeclare module for device instantiation.

# Design Details

The following drivers are adapted to use codegen for device instantiation:
  - I2C: i2c_ll_stm32.c (commit: drivers: i2c: stm32: use codegen driver instantiation)
  - SPI: spi_ll_stm32.c (commit: drivers: spi: stm32: use codegen driver instantiation)
  - SERIAL: uart_stm32.c: (commit: drivers: serial: stm32: use codegen driver instantiation)
  - CAN: stm32_can.c: (commit: drivers: can: stm32: use codegen driver instantiation)
  - PWM: pwm_stm32.c (drivers: pwm: stm32: use codegen driver instantiation)
  - INT: exti_stm32.c (drivers: interrupt: stm32: use codegen driver instantiation)

Codegen #10885  provides two device declaration functions in it's devicedeclare module.
  - devicedeclare.device_declare
    Provides a sophisticated device declaration API. It is transferred from @erwango 's work in #8561.
  - devicedeclare.device_declare_single/multi
    Provides a more raw device declaration API that allows for a fine grained control of the device instantiation template. It is the basis also for devicedeclare.device_declare.

All drivers above except INT are adapted to use the devicedeclare.device_declare_multi API.

The interrupt driver uses the devicedeclare.device_declare API. It is a minimal conversion to show the feasibility. For elaborated examples of the devciedeclare.device_declare API see #8561.

Codegen's device declaration is done using a device info template with placeholders that is used for every device instance. The placeholders are replaced by the respective device tree or device instance values:
  - Device tree property placeholder are defined by the property path; either ${[property path]} for properties of the actual device instance or ${[device id]:[property path]} for another device instance property.
  - The device instance placeholders ${device-[xxx]} provide device instance specific info that is autogenerated eg. device name, driver name, name of the data structure ...

```
/**
 * @code{.codegen}
 * codegen.import_module('devicedeclare')
 *
 * device_configs = ['CONFIG_I2C_{}'.format(x) for x in range(0, 5)]
 * driver_names = ['I2C_{}'.format(x) for x in range(0, 5)]
 * device_inits = 'i2c_stm32_init'
 * device_levels = 'POST_KERNEL'
 * device_prios = 'CONFIG_KERNEL_INIT_PRIORITY_DEVICE'
 * device_api = 'api_funcs'
 * device_info = \
 * """
 * #if CONFIG_I2C_STM32_INTERRUPT
 * DEVICE_DECLARE(${device-name});
 * static void ${device-config-irq}(struct device *dev)
 * {
 * #ifdef CONFIG_I2C_STM32_COMBINED_INTERRUPT
 *         IRQ_CONNECT(${interrupts/combined/irq}, ${interrupts/combined/priority}, \\
 *                     stm32_i2c_combined_isr, \\
 *                     DEVICE_GET(${device-name}), 0);
 *         irq_enable(${interrupts/combined/irq});
 * #else
 *         IRQ_CONNECT(${interrupts/event/irq}, ${interrupts/event/priority}, \\
 *                     stm32_i2c_event_isr, \\
 *                     DEVICE_GET(${device-name}), 0);
 *         irq_enable(${interrupts/event/irq});
 *         IRQ_CONNECT(${interrupts/error/irq}, ${interrupts/error/priority}, \\
 *                     stm32_i2c_error_isr, \\
 *                     DEVICE_GET(${device-name}), 0);
 *         irq_enable(${interrupts/error/irq});
 * #endif
 * }
 * #endif
 * static const struct i2c_stm32_config ${device-config-info} = {
 *         .i2c = (I2C_TypeDef *)${reg/0/address/0},
 *         .pclken.bus = ${clocks/0/bus},
 *         .pclken.enr = ${clocks/0/bits},
 * #if CONFIG_I2C_STM32_INTERRUPT
 *         .irq_config_func = ${device-config-irq},
 * #endif
 *         .bitrate = ${clock-frequency},
 * #if defined(CONFIG_SOC_SERIES_STM32F3X) || defined(CONFIG_SOC_SERIES_STM32F0X)
 *         .ll_clock_source = LL_${driver-name}_CLOCKSOURCE,
 * #endif
 * };
 * static struct i2c_stm32_data ${device-data};
 * """
 *
 * devicedeclare.device_declare_multi( \
 *     device_configs,
 *     driver_names,
 *     device_inits,
 *     device_levels,
 *     device_prios,
 *     device_api,
 *     device_info)
 * @endcode{.codegen}
 */
/** @code{.codeins}@endcode */
```

# Test Strategy

This PR extends the SPI loopback test for the NUCLEO F091RC board. It was tested using this test and some
sample applications.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

### CODE REVIEW

**Comment by SavinayDharmappa** [comment]
@b0661 some drivers of QMSI uses DEVICE_DEFINE() instead of DEVICE_AND_API_INIT() macro  to initialize drivers providing power management support via the PM control function. The codgen by default substitues DEVICE_AND_API_INIT(). Is there any way to specify codegen to substitute DEVICE_DEFINE() macro instead of DEVICE_AND_API_INIT() macro?

**Comment by b0661** [comment]
@SavinayDharmappa 

> Is there any way to specify codegen to substitute DEVICE_DEFINE() macro instead of DEVICE_AND_API_INIT() macro?

Not implemented at the moment. But should not be a problem to switch between DEVICE_DEFINE() and   DEVICE_AND_API_INIT() based on pm_control value.

device_declare_multi() may become

```
def device_declare_multi(device_configs,
                         driver_names,
                         device_inits,
                         device_pm_controls,
                         device_levels,
                         device_prios,
                         device_api,
                         device_info,
                         device_defaults = {}):
```

device_pm_controls maybe a list of pm_control functions or a single pm_control function or None.

File: scripts/codegen/modules/devicedeclare.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    return irq_bootstrap_info
+
+
+def device_declare(compatibles, init_prio_flag, kernel_level, irq_func,
+                   init_func, api, data_struct, config_struct):
+
+    config_struct = _device_generate_struct('config', config_struct)
+    data_struct = _device_generate_struct('data', data_struct)
+    if api is None:
+        api = "(*(const int *)0)"
```
Comment (evenl) [inline scripts/codegen/modules/devicedeclare.py]: Should the template trust that the target toolchain has a sane NULL value?

File: drivers/can/stm32_can.c [hunk 2]
```diff
@@ -841,51 +841,66 @@ static const struct can_driver_api can_api_funcs = {
...
-	can->IER |= CAN_IT_TME | CAN_IT_ERR | CAN_IT_FMP0 | CAN_IT_FMP1;
-}
-
-#endif /*CONFIG_CAN_1*/
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_CAN_1', ]
+ * driver_names = ['CAN_1', ]
```
Comment (erwango) [inline drivers/can/stm32_can.c]: As mentioned several times, this should not be present in API.
This information should be extracted from device tree and drivers <> EDTS binding should be done by 'compatible' property. This is literally its main and only use.
cf #8561

File: drivers/can/stm32_can.c [hunk 3]
```diff
@@ -841,51 +841,66 @@ static const struct can_driver_api can_api_funcs = {
...
+ * device_levels = 'POST_KERNEL'
+ * device_prios = 'CONFIG_KERNEL_INIT_PRIORITY_DEVICE'
+ * device_api = 'can_api_funcs'
+ * device_info = \
+ * """
+ * DEVICE_DECLARE(${device-name});
+ * static void ${device-config-irq}(CAN_TypeDef *can)
+ * {
+ *         LOG_DBG("Enable ${driver-name} IRQ");
+ * #ifdef CONFIG_SOC_SERIES_STM32F0X
```
Comment (erwango) [inline drivers/can/stm32_can.c]: 'compatible' should be used to identify and provide variations of the same driver.
`#ifdef CONFIG_SOC_SERIES_STM32F0X` should not be present in templated code.

File: drivers/can/stm32_can.c [hunk 4]
```diff
@@ -841,51 +841,66 @@ static const struct can_driver_api can_api_funcs = {
...
+ * device_configs = ['CONFIG_CAN_1', ]
+ * driver_names = ['CAN_1', ]
+ * device_inits = 'can_stm32_init'
+ * device_levels = 'POST_KERNEL'
+ * device_prios = 'CONFIG_KERNEL_INIT_PRIORITY_DEVICE'
+ * device_api = 'can_api_funcs'
+ * device_info = \
+ * """
+ * DEVICE_DECLARE(${device-name});
+ * static void ${device-config-irq}(CAN_TypeDef *can)
```
Comment (erwango) [inline drivers/can/stm32_can.c]: This whole piece of code, IRQ configuration functions, should be extracted by API.
API should be able to detect if node specifies IRQs, and generated  this code if needed, cf #8561

File: drivers/i2c/i2c_ll_stm32.c [hunk 5]
```diff
@@ -20,6 +21,23 @@ LOG_MODULE_REGISTER(i2c_ll_stm32);
 
 #include "i2c-priv.h"
 
+#if defined(CONFIG_SOC_SERIES_STM32F3X) || defined(CONFIG_SOC_SERIES_STM32F0X)
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c]: This changes are not related to code generation. Remove from there

File: drivers/i2c/i2c_ll_stm32.c [hunk 6]
```diff
@@ -28,18 +46,24 @@ int i2c_stm32_runtime_configure(struct device *dev, u32_t config)
...
-	LL_RCC_ClocksTypeDef rcc_clocks;
-
-	/*
-	 * STM32F0/3 I2C's independent clock source supports only
-	 * HSI and SYSCLK, not APB1. We force clock variable to
-	 * SYSCLK frequency.
-	 */
-	LL_RCC_GetSystemClocksFreq(&rcc_clocks);
-	clock = rcc_clocks.SYSCLK_Frequency;
+	if (cfg->ll_clock_source != LL_I2C_NO_CLOCKSOURCE) {
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c]: Remove from this change

File: drivers/i2c/i2c_ll_stm32.c [hunk 7]
```diff
@@ -173,32 +197,10 @@ static int i2c_stm32_init(struct device *dev)
 	clock_control_on(clock, (clock_control_subsys_t *) &cfg->pclken);
 
 #if defined(CONFIG_SOC_SERIES_STM32F3X) || defined(CONFIG_SOC_SERIES_STM32F0X)
-	/*
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c]: Remove from this change

File: drivers/spi/spi_ll_stm32.c [hunk 8]
```diff
@@ -338,12 +355,12 @@ static int spi_stm32_configure(struct device *dev,
 	spi_context_cs_configure(&data->ctx);
 
 	LOG_DBG("Installed config %p: freq %uHz (div = %u),"
-		    " mode %u/%u/%u, slave %u",
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Why modifying this?

File: drivers/spi/spi_ll_stm32.c [hunk 9]
```diff
@@ -469,129 +486,60 @@ static int spi_stm32_init(struct device *dev)
...
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_3), 0);
-	irq_enable(CONFIG_SPI_3_IRQ);
-}
-#endif
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_SPI_{}'.format(x) for x in range(1, 7)]
+ * driver_names = ['SPI_{}'.format(x) for x in range(1, 7)]
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Cf comments on CAN driver

**Review by erwango — CHANGES_REQUESTED** [review]
Pleas rebase on topic-EDTS branch.
Same comments as always on the need for higher level device_declare API

**Comment by erwango** [comment]
@b0661 

We've been trying to formalize requirements for driver devices generation here: https://github.com/zephyrproject-rtos/zephyr/issues/10904 (still open for debate, if you want to).

To my view, what is missing to the various PR you've been proposing is adherence to above requirements (which were not written down until now, I hope having a formalized version makes things clearer now). This is why Iv'e been pushing https://github.com/zephyrproject-rtos/zephyr/pull/8561.

Though, I don't want to push this despite you, so if you feel interested, feel free to push your own version of a codegen solution that could work with requirements.

File: scripts/codegen/modules/devicedeclare.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+            '\t\tDEVICE_GET(${device-name}),\n' + \
+            '\t\t0);\n' + \
+            '\tirq_enable({});\n\n'.format(irq_num)
+    irq_bootstrap_info += \
+        '}\n' + \
+        '#endif /* {} */\n\n'.format(irq_flag)
+    return irq_bootstrap_info
+
+
+def device_declare(compatibles, init_prio_flag, kernel_level, irq_func,
```
[the remark this answers is not in the document]
Comment (b0661) [reply scripts/codegen/modules/devicedeclare.py]: @erwango, can you explain. It is your device_declare variant.

File: scripts/codegen/edts.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+        # Try to get a lock for the database file
+        # If we do not get the log for 10 seconds an
+        # exception is thrown.
+        try:
+            with edts_lock.acquire(timeout = 10):
+                if extract:
+                    # File produced by the DT compiler
+                    board = self.cmake_variable("BOARD")
+                    dts_compiled = Path(prj_bin_dir, '{}.dts_compiled'.format(board))
```
[the remark this answers is not in the document]
Comment (b0661) [reply scripts/codegen/edts.py]: The EDTS database class that I am using has it´s own extract function. It does not depend on extract_dts_includes to fill the database. It the file is already available it just loads the json file. You comment on the etxract part if no edts.json file is available.

File: scripts/codegen/cmake.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+#
+
+import os
+import sys
+import re
+from collections import OrderedDict
+from pathlib import Path
+
+
+class CMakeCacheEntry:
```
[the remark this answers is not in the document]
Comment (b0661) [reply scripts/codegen/cmake.py]: For driver writing.

File: drivers/serial/uart_stm32.c [hunk 13]
```diff
@@ -352,161 +353,52 @@ static int uart_stm32_init(struct device *dev)
...
- * so APB1 GRP2 should be accessed instead.
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_UART_STM32_PORT_{}'.format(x) \
+ *                   for x in range(1, 11)] + \
+ *                  ['CONFIG_UART_STM32_LPUART_{}'.format(x) \
+ *                   for x in range(1, 2)]
+ * driver_names = ['UART_{}'.format(x) for x in range(1, 11)] + \
```
[the remark this answers is not in the document]
Comment (b0661) [reply drivers/serial/uart_stm32.c]: > I don't understand why you do this, you already have the label field that you can use?

Later on I will search for device where the label value fits to the ones that I set up here.

> What if the order of devices in the DTS data is not from low to high?

That does not matter. I just search for a device with the given label value.

> Is there any guarantee that the device order will be the same in the DTS database as in the dts file?

This is not necessary - see above.

File: drivers/can/stm32_can.c (hunk 2 above)
Comment (b0661) [reply drivers/can/stm32_can.c]: As mentionend several times to identify a specific driver instance the compatible is not enough. There are usual several driver instantiations for one compatible. So you need something more to identify a single one and use the 'label' property for that. But if you use the 'label' property in addition to the 'compatible' you find out that the 'label' alone is unique und you really do not need the 'compatible'. Taking the 'compatible' and the 'label' is just a waste of computing.

File: drivers/can/stm32_can.c (hunk 3 above)
Comment (b0661) [reply drivers/can/stm32_can.c]: Hey this is a proof of concept and not a full blown driver rework, I expect the CAN driver author to do the rework in the end when there is a code generation facility. My focus was the extraction of device tree data not to make that work without some ifdefs.

File: drivers/spi/spi_ll_stm32.c (hunk 8 above)
Comment (b0661) [reply drivers/spi/spi_ll_stm32.c]: Not by intention. Seems to have happened when SYS_LOG was changed to LOG and I updated.

File: drivers/i2c/i2c_ll_stm32.c (hunk 6 above)
Comment (b0661) [reply drivers/i2c/i2c_ll_stm32.c]: There were compile errors in CI before !???!  Is this gone?

File: drivers/i2c/i2c_ll_stm32.c (hunk 5 above)
Comment (b0661) [reply drivers/i2c/i2c_ll_stm32.c]: In these drivers I added changes whenever CI failed. As already mentioned it is a proof of concept. I do not claim that every driver is right but I want CI to pass. I can not demonstrate the viability of Codegen without CI to pass.

**Comment by b0661** [comment]
@erwango 

Thank you for the comments. I am not quite shure what to do with the comments that relate to driver functions and not to code generation from DTS.

- If codegen does not get merged all this will not be necessary. 
- If codegen is merged the authors of the drivers should do the changes looking at here to have it as a starting point.

My goal for all drivers is to show Codegen works and the driver passes CI and sometimes some other tests.

**Comment by carlescufi** [comment]
@b0661 thanks for the PR. During my discussions with the rest of Zephyr TSC members, most of us expressed a dislike for having the generated code embedded in the actual .c file. Instead we would prefer to have a clean separation between generated code (i.e. instantiation) and the actual driver code.
Is this something that would be possible with codegen? I assume it is, because I believe this came up before, but just double-checking here.

**Comment by b0661** [comment]
> . ...Instead we would prefer to have a clean separation between generated code (i.e. instantiation) and the actual driver code.
Is this something that would be possible with codegen? I assume it is, because I believe this came up before, but just double-checking here.

Yes it is possible. For codegen you simply define the template file and the destination file. It the template file happens to be called my_file.h.in and the destination file my_file.h codegen will produce it the way you are expecting it.

**Comment by b0661** [comment]
@carlescufi

> As mentioned by @erwango this should be rebased on top of the topic branch. 

Will do - but I have to finish some work yet. 

What about the build system integration of codegen (btw. as you asked that: I wrote that - not @SebastianBoe). @anangl somehow hijacked the codegen build system integration - will that be changed?

**Pull request closed without merging**